### PR TITLE
url: improve private IP protection

### LINKED ISSUE #2284: url: private_resolution/dns_resolution useless
### Description
The url.enable_private_resolution and url.enable_dns_resolution settings do not work as advertised, and the concept of the latter is fatally flawed.

The current `url.py` private-address protection logic is as follows:
```python
if not enable_private_resolution:
    if host is ip and ip is private:
        return E_NAUGHTY
    if enable_dns_resolution and resolve(host) is private:
        return E_NAUGHTY
post_title(url)
```
This has many problems, as demonstrated below.

### Reproduction steps
Oh lordy.

#### Scenario 1: Attacker has $2
1. Buy a domain, e.g. `haha.bad`
2. Create a DNS record: `haha.bad A 127.0.0.1`
3. Post a link in IRC: `http://haha.bad/csrf/me/daddy`

If enable_dns_resolution is disabled, this bypasses all private-address protection.

#### Scenario 2: Attacker doesn't have $2
1. Configure webserver to respond to requests with `302 http://127.0.0.1/csrf/me/daddy`
2. Post a link to `http://$webserver_ip/`

The posted link is a public IP, so Sopel will happily load it and follow the redirect to the private address.

#### Scenario 3: Attacker has $2, but the above is fixed
1. Buy a domain, e.g. `haha.bad`
2. Set the nameservers for `haha.bad` to an IP you control
3. Run a script on that IP with the following behavior:
```
attack = False
if attack := not attack:
    return dns_record(type="A", address="1.2.3.4", ttl=0)
return dns_record(type="A", address="127.0.0.1", ttl=0)
```
In the checking stage, Sopel will see the address 1.2.3.4. When performing the request, 127.0.0.1 will be used without validation. ([TOCTOU](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use))

#### Scenario 4: Attacker has $2 and target is IPv6
1. Buy a domain, e.g. `haha.bad`
2. Point `haha.bad` to the target IPv6 address
3. Post a link to `http://haha.bad/csrf/me/daddy`

`dns.resolver.resolve()` does not request AAAA records, so any combination of DNS and IPv6 passes validation.

### Expected behavior
As explained on IRC, this behavior (especially the first two parts) is objectively broken and gives a false sense of security. We should either remove those config options and not pretend, or we should make url.py _unable_ to talk to private IPs.

### Environment
- Sopel `.version`: Since before the name "Sopel" to present.
- Relevant plugins: url.py

### Notes
Ping @dgw and @Exirel, who requested this be an issue instead of IRC comments and a PR.

### LINKED ISSUE #2348: url: Unexpected error on nonexistent domains
The following code line raises an unhandled `dns.resolver.NXDOMAIN` for URLs at nonexistent domains:

https://github.com/sopel-irc/sopel/blob/eff60e9b49a14e1f3c0c4dab5bb4d2ddbf0098a5/sopel/modules/url.py#L377

Some variant of this operation has been in the code for a while, but the context has changed so much during 8.x that I decided not to follow `git blame` all the way back to the first addition of `dns.resolver`. Seemed like a waste of time.

### PR DESCRIPTION
**Pull request by half-duplex**
### Description
Atop #2282
Fixes #2284

Draft implementation should resolve #2348.

- [x] Document TOCTOU

### Checklist
- [x] I have read [CONTRIBUTING.md](https://github.com/sopel-irc/sopel/blob/master/CONTRIBUTING.md)
- [x] I can and do license this contribution under the EFLv2
- [x] No issues are reported by `make qa` (runs `make quality` and `make test`)
- [x] I have tested the functionality of the things this change touches

### CODE REVIEW

**Comment by half-duplex** [comment]
- Passing bot.memory to find_title feels weird, but it's needed for the safety checks, which need to be done each redirect
- Both methods of getting the IPs for a hostname ignore whether a system has working IPv6. This is a really annoying problem I didn't foresee.
  - Check for a global IPv6 address? (But some working v6 doesn't have that)
  - Check for IPv6 routing? (Feels dirty
  - If TTL is over a second or two, assume it'll be cached and accept the edge case TOCTOUs?
    - Needs fallback. Fall back to only supporting IPv4? Gross, but only needed when TTL<2, so would work 99.999% of the time.

**Comment by dgw** [comment]
With #2282 merged, presumably this can proceed toward Ready status after a rebase.

Any idea why this affects VCR tests? Somehow it is causing the example request for google.com to appear as a direct IP instead.

**Comment by half-duplex** [comment]
If the test checks for the address actually contacted, it'll be getting something different from what it used to in some cases. I figured I'd look into that after seeing if I'm even continuing with this redirect handling strategy because of the above questions.

**Comment by dgw** [comment]
Assuming this PR graduates from draft status, please remove `dnspython` from requirements. These changes remove its use from `url.py` (🥳), and nothing else uses it.

**Comment by half-duplex** [comment]
Rebased, but still looking for feedback on the above questions.

File: sopel/modules/url.py [hunk 1]
```diff
@@ -436,32 +419,107 @@ def check_callbacks(bot: SopelWrapper, url: str, use_excludes: bool = True) -> b
     )
 
 
-def find_title(url: str, verify: bool = True) -> Optional[str]:
-    """Return the title for the given URL.
+def find_title(
+    url: str,
+    verify: bool = True,
+    allow_local: bool = False,
+    memory: tools.SopelMemory = {},
```
Comment (Exirel) [inline sopel/modules/url.py:426]: I'd replace that by 2 parameters:

* `unsafe_urls`: a list of unsafe URLs
* `unsafe_domains`: a list of unsafe domain name

You would then need to call the function as follow:

```python
unsafe_urls = [
    unsafe_url
    for unsafe_url, safe_data in bot.memory.get("safety_cache", {}).items()
    if safe_data.get('positives', 0) > 0
]
unsafe_domains = bot.memory.get("safety_cache_local", {}).values()

result = find_title(url, verify, allow_local, unsafe_urls, unsafe_domains)
```

No need for a SopelMemory anymore, no need to pass the bot around, etc.

File: sopel/modules/url.py [hunk 2]
```diff
@@ -436,32 +419,107 @@ def check_callbacks(bot: SopelWrapper, url: str, use_excludes: bool = True) -> b
...
+    session.mount("https://", HostAdapter())
+    session.headers = dict(DEFAULT_HEADERS)
+    while redirects_left > 0:
+        redirects_left -= 1
+        parsed_url = urlparse(url)
+
+        # Avoid fetching known malicious links
+        safety_cache = bot.memory.get("safety_cache", {})
+        safety_cache_local = bot.memory.get("safety_cache_local", {})
+        if url in safety_cache and safety_cache[url]["positives"] > 0:
```
Comment (Exirel) [inline sopel/modules/url.py:447]: Note that you are using `bot.memory` but your argument is `memory`.

File: sopel/modules/url.py [hunk 3]
```diff
@@ -436,32 +419,107 @@ def check_callbacks(bot: SopelWrapper, url: str, use_excludes: bool = True) -> b
...
+                LOGGER.debug("Ignoring unsafe URL: %r", url)
+                return None
```
Comment (Exirel) [inline sopel/modules/url.py:449]:
```suggestion
            LOGGER.debug("Ignoring unsafe URL: %r", url)
            return None
```

**Review by Exirel — CHANGES_REQUESTED** [review]

**Comment by dgw** [comment]
@half-duplex Did you still want to get this in? Haven't seen any activity for a while.

**Comment by half-duplex** [comment]
For when I look at this next: https://github.com/requests/toolbelt/pull/293 was merged

File: sopel/modules/url.py (hunk 1 above)
Comment (half-duplex) [reply sopel/modules/url.py:426]: Good point, thanks.

**Comment by half-duplex** [comment]
Some dns tests:
http://lo.4.sopel.sec.gd/
http://net.4.sopel.sec.gd/
http://net2.4.sopel.sec.gd/
http://lo.6.sopel.sec.gd/
http://lo4.6.sopel.sec.gd/
http://ula.6.sopel.sec.gd/
http://ll.6.sopel.sec.gd/

**Comment by dgw** [comment]
Since @half-duplex reminded us this exists during another (unrelated) IRC discussion today, I came by to give it a look. Nothing jumps out at me, but I will defer final approval to @Exirel as the original reviewer who requested changes.

I'm excited at getting rid of `dnspython` as a dependency, though. 😁

**Comment by dgw** [comment]
> I'm excited at getting rid of `dnspython` as a dependency, though. 😁

Even more so now that `dnspython` 2.7.0 has broken type-checks on `master` (see https://github.com/sopel-irc/sopel/pull/2594#issuecomment-2398040528). Python 3.8 isn't supported in the new version, so that CI job proceeded past the `lint` step to `test` before getting canceled when a sibling failed. I also confirmed myself, `make lint` succeeds on `master` with `dnspython` 2.6 and fails with 2.7.

@Exirel when you've some time to look at this again, it'd be much appreciated 😸

**Review by Exirel — APPROVED** [review]
Good riddance @ dnspython!

**Pull request merged**